File: CastIt/Services/CastItHubClientService.cs
using CastIt.Application.Server;
using CastIt.Domain.Dtos.Requests;
using CastIt.Domain.Dtos.Responses;
using CastIt.Domain.Enums;
using CastIt.Domain.Interfaces;
using CastIt.Domain.Models.Device;
using CastIt.Infrastructure.Models;
using CastIt.Interfaces;
using Microsoft.AspNetCore.SignalR.Client;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace CastIt.Services
{
    public class CastItHubClientService : ICastItHubClientService
    {
        private readonly ILogger<CastItHubClientService> _logger;
        private HubConnection _connection;

        public event Action OnClientConnected;
        public event Action OnClientDisconnected;

        public event Action<ServerPlayerStatusResponseDto> OnPlayerStatusChanged;
        public event Action<ServerAppSettings> OnPlayerSettingsChanged;
        public event Action<IReceiver> OnCastDeviceSet;
        public event Action<List<IReceiver>> OnCastDevicesChanged;
        public event Action OnCastDeviceDisconnected;
        public event Action<AppMessageType> OnServerMessage;
        public event Action OnStoppedPlayback;

        public event Action<List<GetAllPlayListResponseDto>> OnPlayListsLoaded;
        public event Action<GetAllPlayListResponseDto> OnPlayListAdded;
        public event Action<GetAllPlayListResponseDto> OnPlayListChanged;
        public event Action<List<GetAllPlayListResponseDto>> OnPlayListsChanged;
        public event Action<long> OnPlayListDeleted;
        public event Action<long, bool> OnPlayListBusy;

        public event Action<FileItemResponseDto> OnFileAdded;
        public event Action<FileItemResponseDto> OnFileChanged;
        public event Action<List<FileItemResponseDto>> OnFilesChanged;
        public event Action<long, long> OnFileDeleted;
        public event Action<FileItemResponseDto> OnFileLoading;
        public event Action<FileItemResponseDto> OnFileLoaded;
        public event Action<FileItemResponseDto> OnFileEndReached;

        public string IpAddress { get; set; }

        public CastItHubClientService(ILogger<CastItHubClientService> logger)
        {
            _logger = logger;
        }

        public async Task<bool> Init()
        {
            //Kinda hack...
            bool serverStarted = await StartServerIfNotStarted();
            if (!serverStarted)
                return false;

            bool connectionBuilt = await BuildConnection();
            if (!connectionBuilt)
                return false;

            ListenForEvents();
            var connected = await ConnectToHub();
            if (connected)
                OnClientConnected?.Invoke();

            return connected;
        }

        public async Task<bool> Init(string serverIpAddress)
        {
            if (string.IsNullOrWhiteSpace(serverIpAddress) || !Uri.TryCreate(serverIpAddress, UriKind.Absolute, out _))
            {
                throw new ArgumentOutOfRangeException(nameof(serverIpAddress), serverIpAddress, "You need to provide a valid server ip address");
            }

            bool connectionBuilt = await BuildConnection(serverIpAddress, 1);
            if (!connectionBuilt)
                return false;

            ListenForEvents();
            var connected = await ConnectToHub(1);
            if (connected)
                OnClientConnected?.Invoke();

            return connected;
        }

        public async Task DisposeAsync()
        {
            OnClientDisconnected = null;
            OnClientConnected = null;
            OnPlayerStatusChanged = null;
            OnPlayerSettingsChanged = null;
            OnCastDeviceSet = null;
            OnCastDevicesChanged = null;
            OnServerMessage = null;
            OnStoppedPlayback = null;
            OnPlayListsLoaded = null;
            OnPlayListAdded = null;
            OnPlayListChanged = null;
            OnPlayListsChanged = null;
            OnPlayListDeleted = null;
            OnPlayListBusy = null;
            OnFileAdded = null;
            OnFileChanged = null;
            OnFilesChanged = null;
            OnFileDeleted = null;
            OnFileLoading = null;
            OnFileLoaded = null;
            OnFileEndReached = null;
            if (_connection != null)
            {
                await _connection.DisposeAsync();
            }
        }

        #region Player methods
        public async Task Play(long playListId, long fileId, bool force, bool fileOptionsChanged = false)
        {
            try
            {
                var request = new PlayFileRequestDto
                {
                    Id = fileId,
                    PlayListId = playListId,
                    Force = force,
                    FileOptionsChanged = fileOptionsChanged
                };
                await _connection.InvokeAsync(nameof(Play), request);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(Play)}: Unknown error occurred");
            }
        }

        public async Task GoToSeconds(double seconds)
        {
            try
            {
                await _connection.InvokeAsync(nameof(GoToSeconds), seconds);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(GoToSeconds)}: Unknown error occurred");
            }
        }

        public async Task SkipSeconds(double seconds)
        {
            try
            {
                await _connection.InvokeAsync(nameof(SkipSeconds), seconds);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(SkipSeconds)}: Unknown error occurred");
            }
        }

        public async Task GoTo(bool next, bool previous)
        {
            try
            {
                await _connection.InvokeAsync(nameof(GoTo), next, previous);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(GoTo)}: Unknown error occurred");
            }
        }

        public async Task TogglePlayBack()
        {
            try
            {
                await _connection.InvokeAsync(nameof(TogglePlayBack));
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(TogglePlayBack)}: Unknown error occurred");
            }
        }

        public async Task StopPlayBack()
        {
            try
            {
                await _connection.InvokeAsync(nameof(StopPlayBack));
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(StopPlayBack)}: Unknown error occurred");
            }
        }

        public async Task SetVolume(double level, bool isMuted)
        {
            var request = new SetVolumeRequestDto
            {
                IsMuted = isMuted,
                VolumeLevel = level
            };
            try
            {
                await _connection.InvokeAsync(nameof(SetVolume), request);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(SetVolume)}: Unknown error occurred");
            }
        }

        public async Task UpdateSettings(ServerAppSettings settings)
        {
            try
            {
                await _connection.InvokeAsync(nameof(UpdateSettings), settings);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(UpdateSettings)}: Unknown error occurred");
            }
        }

        public async Task ConnectToCastDevice(string id)
        {
            try
            {
                await _connection.InvokeAsync(nameof(ConnectToCastDevice), id);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(ConnectToCastDevice)}: Unknown error occurred");
            }
        }

        public async Task RefreshCastDevices(TimeSpan ts)
        {
            try
            {
                await _connection.InvokeAsync(nameof(RefreshCastDevices), ts);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(RefreshCastDevices)}: Unknown error occurred");
            }
        }

        public async Task SetFileSubtitlesFromPath(string path)
        {
            try
            {
                await _connection.InvokeAsync(nameof(SetFileSubtitlesFromPath), path);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(SetFileSubtitlesFromPath)}: Unknown error occurred");
            }
        }
        #endregion

        #region PlayList Methods

        public async Task<PlayListItemResponseDto> AddNewPlayList()
        {
            try
            {
                return await _connection.InvokeAsync<PlayListItemResponseDto>(nameof(AddNewPlayList));
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(AddNewPlayList)}: Unknown error occurred");
                throw;
            }
        }

        public async Task<PlayListItemResponseDto> GetPlayList(long id)
        {
            try
            {
                return await _connection.InvokeAsync<PlayListItemResponseDto>(nameof(GetPlayList), id);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(GetPlayList)}: Unknown error occurred");
                throw;
            }
        }

        public async Task UpdatePlayList(long id, string name)
        {
            var request = new UpdatePlayListRequestDto
            {
                Name = name,
            };
            try
            {
                await _connection.InvokeAsync(nameof(UpdatePlayList), id, request);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(UpdatePlayList)}: Unknown error occurred");
            }
        }

        public async Task UpdatePlayListPosition(long playListId, int newIndex)
        {
            try
            {
                await _connection.InvokeAsync(nameof(UpdatePlayListPosition), playListId, newIndex);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(UpdatePlayListPosition)}: Unknown error occurred");
            }
        }

        public async Task SetPlayListOptions(long playListId, bool loop, bool shuffle)
        {
            var request = new SetPlayListOptionsRequestDto
            {
                Loop = loop,
                Shuffle = shuffle
            };
            try
            {
                await _connection.InvokeAsync(nameof(SetPlayListOptions), playListId, request);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(SetPlayListOptions)}: Unknown error occurred");
            }
        }

        public async Task DeletePlayList(long id)
        {
            try
            {
                await _connection.InvokeAsync(nameof(DeletePlayList), id);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(DeletePlayList)}: Unknown error occurred");
            }
        }

        public async Task DeleteAllPlayLists(long exceptId)
        {
            try
            {
                await _connection.InvokeAsync(nameof(DeleteAllPlayLists), exceptId);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(DeleteAllPlayLists)}: Unknown error occurred");
            }
        }

        public async Task RemoveFiles(long playListId, List<long> ids)
        {
            try
            {
                await _connection.InvokeAsync(nameof(RemoveFiles), playListId, ids);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(RemoveFiles)}: Unknown error occurred");
            }
        }

        public async Task RemoveFilesThatStartsWith(long playListId, string path)
        {
            try
            {
                await _connection.InvokeAsync(nameof(RemoveFilesThatStartsWith), playListId, path);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(RemoveFilesThatStartsWith)}: Unknown error occurred");
            }
        }

        public async Task RemoveAllMissingFiles(long playListId)
        {
            try
            {
                await _connection.InvokeAsync(nameof(RemoveAllMissingFiles), playListId);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(RemoveAllMissingFiles)}: Unknown error occurred");
            }
        }

        public async Task AddFolders(long playListId, bool includeSubFolder, params string[] folders)
        {
            var request = new AddFolderOrFilesToPlayListRequestDto
            {
                Folders = folders.ToList(),
                IncludeSubFolders = includeSubFolder
            };

            try
            {
                await _connection.InvokeAsync(nameof(AddFolders), playListId, request);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(AddFolders)}: Unknown error occurred");
            }
        }

        public async Task AddFiles(long playListId, params string[] files)
        {
            var request = new AddFolderOrFilesToPlayListRequestDto
            {
                Files = files.ToList(),
            };

            try
            {
                await _connection.InvokeAsync(nameof(AddFiles), playListId, request);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(AddFiles)}: Unknown error occurred");
            }
        }

        public async Task AddUrlFile(long playListId, string url, bool onlyVideo)
        {
            var request = new AddUrlToPlayListRequestDto
            {
                Url = url,
                OnlyVideo = onlyVideo
            };

            try
            {
                await _connection.InvokeAsync(nameof(AddUrlFile), playListId, request);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(AddUrlFile)}: Unknown error occurred");
            }
        }

        public async Task SortFiles(long playListId, SortModeType sortMode)
        {
            try
            {
                await _connection.InvokeAsync(nameof(SortFiles), playListId, sortMode);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(SortFiles)}: Unknown error occurred");
            }
        }
        #endregion

        #region File Methods
        public async Task LoopFile(long playListId, long id, bool loop)
        {
            try
            {
                await _connection.InvokeAsync(nameof(LoopFile), playListId, id, loop);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(LoopFile)}: Unknown error occurred");
            }
        }

        public async Task DeleteFile(long playListId, long id)
        {
            try
            {
                await _connection.InvokeAsync(nameof(DeleteFile), playListId, id);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(DeleteFile)}: Unknown error occurred");
            }
        }

        public async Task SetFileOptions(int streamIndex, bool isAudio, bool isSubTitle, bool isQuality)
        {
            var request = new SetFileOptionsRequestDto
            {
                StreamIndex = streamIndex,
                IsAudio = isAudio,
                IsSubTitle = isSubTitle,
                IsQuality = isQuality,
            };
            try
            {
                await _connection.InvokeAsync(nameof(SetFileOptions), request);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(SetFileOptions)}: Unknown error occurred");
            }
        }

        public async Task UpdateFilePosition(long playListId, long id, int newIndex)
        {
            try
            {
                await _connection.InvokeAsync(nameof(UpdateFilePosition), playListId, id, newIndex);
            }
            catch (Exception e)
            {
                _logger.LogError(e, $"{nameof(UpdateFilePosition)}: Unknown error occurred");
            }
        }
        #endregion

        #region Helpers
        //TODO: ONCE THIS GETS IMPLEMENTED, REPLACE THIS CODE
        //https://github.com/dotnet/aspnetcore/issues/32534
        public void ListenForEvents()
        {
            //Remember that all the T on On<T> has to be a class with an empty constructor
            ListenPlayerEvents();

            ListenPlayListsEvents();

            ListenFileEvents();
        }

        public async Task<bool> StartServerIfNotStarted()
        {
            int retries = 3;
            while (retries > 0)
            {
                if (WebServerUtils.IsServerAlive())
                    return true;

                _logger.LogInformation($"{nameof(StartServerIfNotStarted)}: Server has not been started... Starting it with retries left = {retries}...");
                bool started = WebServerUtils.StartServer();
                if (started)
                    return true;
                _logger.LogWarning($"{nameof(StartServerIfNotStarted)}: Server could not be started");
                await Task.Delay(1000);
                retries--;
            }

            return false;
        }

        public async Task<bool> BuildConnection(string ipAddress = null, int retries = 3)
        {
            while (retries > 0)
            {
                _logger.LogInformation($"{nameof(BuildConnection)}: Retrieving server's ip address with retries = {retries}...");
                if (string.IsNullOrWhiteSpace(ipAddress))
                {
                    ipAddress = WebServerUtils.GetWebServerIpAddress();
                }
                else
                {
                    _logger.LogInformation($"{nameof(BuildConnection)}: Ip address = {ipAddress} was provided, using that one to try to connect to the server's hub...");
                }

                if (!string.IsNullOrWhiteSpace(ipAddress))
                {
                    if (ipAddress.EndsWith("/"))
                    {
                        ipAddress = ipAddress[..^1];
                    }

                    if (_connection != null)
                    {
                        _connection.Closed -= ConnectionClosed;
                        await _connection.DisposeAsync();
                    }

                    //TODO: MOVE THE CastItHub TO A CONSTANT 
                    var url = $"{ipAddress}/CastItHub";
                    _connection = new HubConnectionBuilder()
                        .WithUrl(url)
                        .Build();

                    _connection.Closed += ConnectionClosed;
                    _logger.LogInformation($"{nameof(BuildConnection)}: The server's hub address is = {url}");
                    IpAddress = ipAddress;
                    return true;
                }

                await Task.Delay(1000);
                retries--;
            }

            return false;
        }

        public async Task<bool> ConnectToHub(int retries = 5)
        {
            while (retries > 0)
            {
                try
                {
                    _logger.LogInformation($"{nameof(ConnectToHub)}: Trying to connect to server's hub with retries = {retries}...");
                    await _connection.StartAsync();
                    _logger.LogInformation($"{nameof(ConnectToHub)}: Connection was successfully established");
                    return true;
                }
                catch (Exception e)
                {
                    _logger.LogError(e, $"{nameof(ConnectToHub)}: Unknown error occurred");
                }

                await Task.Delay(500);
                retries--;
            }
            return false;
        }

        private void ListenPlayerEvents()
        {
            _connection.On<ServerPlayerStatusResponseDto>("PlayerStatusChanged", status => OnPlayerStatusChanged?.Invoke(status));

            _connection.On<ServerAppSettings>("PlayerSettingsChanged", settings => OnPlayerSettingsChanged?.Invoke(settings));

            _connection.On<AppMessageType>("ServerMessage", msg => OnServerMessage?.Invoke(msg));

            _connection.On<Receiver>("CastDeviceSet", device => OnCastDeviceSet?.Invoke(device));

            _connection.On<List<Receiver>>("CastDevicesChanged", devices => OnCastDevicesChanged?.Invoke(devices.Cast<IReceiver>().ToList()));

            _connection.On("CastDeviceDisconnected", () => OnCastDeviceDisconnected?.Invoke());

            _connection.On("StoppedPlayBack", () => OnStoppedPlayback?.Invoke());
        }

        private void ListenPlayListsEvents()
        {
            _connection.On<List<GetAllPlayListResponseDto>>("SendPlayLists", playLists => OnPlayListsLoaded?.Invoke(playLists));

            _connection.On<GetAllPlayListResponseDto>("PlayListAdded", playList => OnPlayListAdded?.Invoke(playList));

            _connection.On<List<GetAllPlayListResponseDto>>("PlayListsChanged", playLists => OnPlayListsChanged?.Invoke(playLists));

            _connection.On<GetAllPlayListResponseDto>("PlayListChanged", playList => OnPlayListChanged?.Invoke(playList));

            _connection.On<long>("PlayListDeleted", id => OnPlayListDeleted?.Invoke(id));

            _connection.On<long, bool>("PlayListIsBusy", (id, isBusy) => OnPlayListBusy?.Invoke(id, isBusy));
        }

        private void ListenFileEvents()
        {
            _connection.On<FileItemResponseDto>("FileAdded", file => OnFileAdded?.Invoke(file));

            _connection.On<FileItemResponseDto>("FileChanged", file => OnFileChanged?.Invoke(file));

            _connection.On<List<FileItemResponseDto>>("FilesChanged", files => OnFilesChanged?.Invoke(files));

            _connection.On<long, long>("FileDeleted", (playListId, id) => OnFileDeleted?.Invoke(playListId, id));

            _connection.On<FileItemResponseDto>("FileLoading", file => OnFileLoading?.Invoke(file));

            _connection.On<FileItemResponseDto>("FileLoaded", file => OnFileLoaded?.Invoke(file));

            _connection.On<FileItemResponseDto>("FileEndReached", file => OnFileEndReached?.Invoke(file));
        }

        private Task ConnectionClosed(Exception ex)
        {
            if (ex != null)
            {
                _logger.LogError(ex, $"{nameof(ConnectionClosed)}: Connection closed");
            }
            OnClientDisconnected?.Invoke();
            return Task.CompletedTask;
        }
        #endregion
    }
}

File: CastIt.Client/lib/infrastructure/castit_hub_client_service.dart
import 'dart:async';

import 'package:castit/domain/enums/enums.dart';
import 'package:castit/domain/models/models.dart';
import 'package:castit/domain/services/castit_hub_client_service.dart';
import 'package:castit/domain/services/logging_service.dart';
import 'package:castit/domain/services/settings_service.dart';
import 'package:signalr_core/signalr_core.dart';
import 'package:tuple/tuple.dart';

const String _sendPlayLists = 'SendPlayLists';
const String _stoppedPlayback = 'StoppedPlayBack';
const String _playListAdded = 'PlayListAdded';
const String _playListChanged = 'PlayListChanged';
const String _playListsChanged = 'PlayListsChanged';
const String _playListDeleted = 'PlayListDeleted';
const String _playListBusy = 'PlayListIsBusy';
const String _fileAdded = 'FileAdded';
const String _fileChanged = 'FileChanged';
const String _filesChanged = 'FilesChanged';
const String _fileDeleted = 'FileDeleted';
const String _fileLoading = 'FileLoading';
const String _fileLoaded = 'FileLoaded';
const String _fileEndReached = 'FileEndReached';
const String _playerStatusChanged = 'PlayerStatusChanged';
const String _playerSettingsChanged = 'PlayerSettingsChanged';
const String _serverMessage = 'ServerMessage';
const String _castDeviceSet = 'CastDeviceSet';
const String _castDevicesChanged = 'CastDevicesChanged';
const String _castDeviceDisconnected = 'CastDeviceDisconnected';

class CastItHubClientServiceImpl implements CastItHubClientService {
  @override
  final StreamController<void> connected = StreamController.broadcast();

  @override
  final StreamController<void> fileLoading = StreamController.broadcast();

  @override
  final StreamController<PlayedFile> fileLoaded = StreamController.broadcast();

  @override
  final StreamController<String> fileLoadingError = StreamController.broadcast();

  @override
  final StreamController<double> fileTimeChanged = StreamController.broadcast();

  @override
  final StreamController<void> filePaused = StreamController.broadcast();

  @override
  final StreamController<void> fileEndReached = StreamController.broadcast();

  @override
  final StreamController<void> disconnected = StreamController.broadcast();

  @override
  final StreamController<void> appClosing = StreamController.broadcast();

  @override
  final StreamController<ServerAppSettings> settingsChanged = StreamController.broadcast();

  @override
  final StreamController<PlayListItemResponseDto?> playlistLoaded = StreamController.broadcast();

  @override
  final StreamController<List<FileItemOptionsResponseDto>> fileOptionsLoaded = StreamController.broadcast();

  @override
  final StreamController<VolumeLevelChangedResponseDto> volumeLevelChanged = StreamController.broadcast();

  @override
  final StreamController<RefreshPlayListResponseDto> refreshPlayList = StreamController.broadcast();

  @override
  final StreamController<AppMessageType> serverMessageReceived = StreamController.broadcast();

  @override
  final StreamController<GetAllPlayListResponseDto> playListAdded = StreamController.broadcast();

  @override
  final StreamController<List<GetAllPlayListResponseDto>> playListsChanged = StreamController.broadcast();

  @override
  final StreamController<Tuple2<bool, GetAllPlayListResponseDto>> playListChanged = StreamController.broadcast();

  @override
  final StreamController<int> playListDeleted = StreamController.broadcast();

  @override
  final StreamController<FileItemResponseDto> fileAdded = StreamController.broadcast();

  @override
  final StreamController<Tuple2<bool, FileItemResponseDto>> fileChanged = StreamController.broadcast();

  @override
  final StreamController<List<FileItemResponseDto>> filesChanged = StreamController.broadcast();

  @override
  final StreamController<Tuple2<int, int>> fileDeleted = StreamController.broadcast();

  final LoggingService _logger;
  final SettingsService _settings;
  HubConnection? _connection;

  @override
  bool get isConnected => _connection?.state == HubConnectionState.connected;

  CastItHubClientServiceImpl(this._logger, this._settings);

  @override
  Future<void> connectToHub() async {
    final url = '${_settings.castItUrl}/castithub';
    try {
      _logger.info(runtimeType, 'Trying to connect to hub on = $url ...');
      final options = HttpConnectionOptions(
        logging: (level, message) {
          switch (level) {
            case LogLevel.trace:
            case LogLevel.debug:
            case LogLevel.none:
            case LogLevel.information:
              // _logger.info(runtimeType, message);
              break;
            case LogLevel.warning:
              // _logger.warning(runtimeType, message);
              break;
            case LogLevel.error:
            case LogLevel.critical:
              _logger.error(runtimeType, message);
          }
        },
      );

      await disconnectFromHub(triggerEvent: false);

      _connection = HubConnectionBuilder().withUrl(url, options).build();
      _listenHubEvents(_connection!);
      await _connection!.start();
    } catch (e, s) {
      _logger.error(runtimeType, '_connectToHub: Error while trying to connect to hub = $url', e, s);
      await disconnectFromHub();
    }
  }

  @override
  Future<void> dispose() async {
    await disconnectFromHub();
    await Future.wait([
      connected.close(),
      fileLoading.close(),
      fileLoaded.close(),
      fileLoadingError.close(),
      fileTimeChanged.close(),
      filePaused.close(),
      fileEndReached.close(),
      disconnected.close(),
      appClosing.close(),
      settingsChanged.close(),
      playListsChanged.close(),
      playlistLoaded.close(),
      fileOptionsLoaded.close(),
      volumeLevelChanged.close(),
      refreshPlayList.close(),
      serverMessageReceived.close(),
    ]);
  }

  @override
  Future<void> disconnectFromHub({bool triggerEvent = true}) async {
    if (_connection != null) {
      _connection!.off(_sendPlayLists);

      _connection!.off(_playListAdded);

      _connection!.off(_playListChanged);

      _connection!.off(_playListsChanged);

      _connection!.off(_playListDeleted);

      _connection!.off(_playListBusy);

      _connection!.off(_fileAdded);

      _connection!.off(_fileChanged);

      _connection!.off(_filesChanged);

      _connection!.off(_fileDeleted);

      _connection!.off(_fileLoading);

      _connection!.off(_fileLoaded);

      _connection!.off(_fileEndReached);

      _connection!.off(_playerStatusChanged);

      _connection!.off(_playerSettingsChanged);

      _connection!.off(_serverMessage);

      _connection!.off(_castDeviceSet);

      _connection!.off(_castDevicesChanged);

      _connection!.off(_castDeviceDisconnected);

      await _connection!.stop();
    }

    _connection = null;
    if (triggerEvent) {
      disconnected.add(null);
    }
  }

  void _listenHubEvents(HubConnection connection) {
    connection.onclose((exception) async {
      await _onHubErrorOrDone(exception);
    });
    connection.on(_sendPlayLists, (message) => _handleHubMsg(_sendPlayLists, message?.first));

    //TODO: IT SEEMS METHODS WITHOUT PARAMS FAILS WITH THIS LIB
    // connection.on(_stoppedPlayback, (message) => _handleHubMsg(_stoppedPlayback, message?.first));

    connection.on(_playListAdded, (message) => _handleHubMsg(_playListAdded, message?.first));

    connection.on(_playListChanged, (message) => _handleHubMsg(_playListChanged, message?.first));

    connection.on(_playListsChanged, (message) => _handleHubMsg(_playListsChanged, message?.first));

    connection.on(_playListDeleted, (message) => _handleHubMsg(_playListDeleted, message?.first));

    connection.on(_playListBusy, (message) => _handleHubMsg(_playListBusy, message?.first));

    connection.on(_fileAdded, (message) => _handleHubMsg(_fileAdded, message?.first));

    connection.on(_fileChanged, (message) => _handleHubMsg(_fileChanged, message?.first));

    connection.on(_filesChanged, (message) => _handleHubMsg(_filesChanged, message?.first));

    connection.on(_fileDeleted, (message) => _handleHubMsg(_fileDeleted, message));

    connection.on(_fileLoading, (message) => _handleHubMsg(_fileLoading, message?.first));

    connection.on(_fileLoaded, (message) => _handleHubMsg(_fileLoaded, message?.first));

    connection.on(_fileEndReached, (message) => _handleHubMsg(_fileEndReached, message?.first));

    connection.on(_playerStatusChanged, (message) => _handleHubMsg(_playerStatusChanged, message?.first));

    connection.on(_playerSettingsChanged, (message) => _handleHubMsg(_playerSettingsChanged, message?.first));

    connection.on(_serverMessage, (message) => _handleHubMsg(_serverMessage, message?.first));

    connection.on(_castDeviceSet, (message) => _handleHubMsg(_castDeviceSet, message?.first));

    connection.on(_castDevicesChanged, (message) => _handleHubMsg(_castDevicesChanged, message?.first));

    connection.on(_castDeviceDisconnected, (message) => _handleHubMsg(_castDeviceDisconnected, message?.first));
  }

  void _handleHubMsg(String msgType, dynamic message) {
    if (msgType != _playerStatusChanged) {
      _logger.info(runtimeType, 'Handling msgType = $msgType');
    }
    switch (msgType) {
      case _sendPlayLists:
        final pls = message as List<dynamic>;
        final playlists = pls.map((e) => GetAllPlayListResponseDto.fromJson(e as Map<String, dynamic>)).toList();
        playListsChanged.add(playlists);
      case _playerSettingsChanged:
        _logger.info(runtimeType, '_handleSocketMsg: Settings were loaded');
        final settings = ServerAppSettings.fromJson(message as Map<String, dynamic>);
        connected.add(null);
        settingsChanged.add(settings);
      case _playerStatusChanged:
        final status = ServerPlayerStatusResponseDto.fromJson(message as Map<String, dynamic>);
        if (status.player.isPaused) {
          filePaused.add(null);
        }
        if (status.playedFile != null) {
          playListChanged.add(Tuple2(true, status.playList!));
          final f = PlayedFile.from(status);
          fileLoaded.add(f);
          fileChanged.add(Tuple2(true, status.playedFile!));
          fileOptionsLoaded.add(status.playedFile!.streams);
          fileTimeChanged.add(f.currentSeconds);
        } else {
          fileEndReached.add(null);
        }
        final volume = VolumeLevelChangedResponseDto(isMuted: status.player.isMuted, volumeLevel: status.player.volumeLevel);
        volumeLevelChanged.add(volume);
      case _playListAdded:
        final newPl = GetAllPlayListResponseDto.fromJson(message as Map<String, dynamic>);
        playListAdded.add(newPl);
      case _playListChanged:
        final updatedPl = GetAllPlayListResponseDto.fromJson(message as Map<String, dynamic>);
        playListChanged.add(Tuple2(false, updatedPl));
      case _playListsChanged:
        final playLists = (message as List<dynamic>).map((e) => GetAllPlayListResponseDto.fromJson(e as Map<String, dynamic>)).toList();
        playListsChanged.add(playLists);
      case _playListDeleted:
        final id = message as int;
        playListDeleted.add(id);
      case _playListBusy:
        break;
      case _fileAdded:
        final file = FileItemResponseDto.fromJson(message as Map<String, dynamic>);
        fileAdded.add(file);
      case _fileChanged:
        final file = FileItemResponseDto.fromJson(message as Map<String, dynamic>);
        fileChanged.add(Tuple2(false, file));
      case _filesChanged:
        final files = (message as List<dynamic>).map((e) => FileItemResponseDto.fromJson(e as Map<String, dynamic>)).toList();
        filesChanged.add(files);
      case _fileDeleted:
        final items = message as List<dynamic>;
        final playListId = items.first as int;
        final fileId = items.last as int;
        fileDeleted.add(Tuple2(playListId, fileId));
      case _fileLoading:
        _logger.info(runtimeType, '_handleSocketMsg: A file is loading...');
        fileLoading.add(null);
      case _serverMessage:
        final code = message as int;
        serverMessageReceived.add(getAppMessageType(code));
      case _stoppedPlayback:
      case _fileEndReached:
        _logger.info(runtimeType, '_handleSocketMsg: File end reached');
        fileEndReached.add(null);
      case _castDevicesChanged:
        break;
      case _castDeviceSet:
        break;
      case _castDeviceDisconnected:
        break;
      case _fileLoaded:
        break;
      default:
        _logger.warning(runtimeType, '_handleSocketMsg: Msg = $msgType is not being handled');
        break;
    }
  }

  Future<void> _onHubErrorOrDone(Exception? exception) async {
    if (exception != null) {
      _logger.error(runtimeType, '_listenHubEvents: Channel error method was called', exception);
    } else {
      _logger.info(runtimeType, '_listenHubEvents: Channel done method was called');
    }
    if (!isConnected) {
      await disconnectFromHub();
    }
  }

  Future<dynamic> _invokeHubMethod(String methodName, List<dynamic> args) async {
    try {
      _logger.info(runtimeType, '_invokeHubMethod: Trying to call method  = $methodName');
      if (_connection == null) {
        _logger.info(runtimeType, '_invokeHubMethod: Connection is null, trying to establish connection before calling method = $methodName');
        await connectToHub();
      }

      if (!isConnected) {
        _logger.warning(runtimeType, '_invokeHubMethod: Cannot invoke method  = $methodName cause the server is not running');
        await disconnectFromHub();
        return;
      }

      return await _connection!.invoke(methodName, args: args);
    } catch (e, s) {
      await disconnectFromHub();
      _logger.error(runtimeType, '_invokeHubMethod: Unknown error', e, s);
    }
    return null;
  }

  @override
  Future<void> playFile(int id, int playListId, {bool force = false}) async {
    final dto = PlayFileRequestDto(id: id, playListId: playListId, force: force);
    await _invokeHubMethod('Play', [dto]);
  }

  @override
  Future<void> gotoSeconds(double seconds) async {
    await _invokeHubMethod('GoToSeconds', [seconds]);
  }

  @override
  Future<void> skipSeconds(double seconds) async {
    await _invokeHubMethod('SkipSeconds', [seconds]);
  }

  @override
  Future<void> goTo({bool next = false, bool previous = false}) async {
    await _invokeHubMethod('GoTo', [next, previous]);
  }

  @override
  Future<void> togglePlayBack() async {
    await _invokeHubMethod('TogglePlayBack', []);
  }

  @override
  Future<void> stopPlayBack() async {
    await _invokeHubMethod('StopPlayback', []);
  }

  @override
  Future<void> setPlayListOptions(int id, {bool loop = false, bool shuffle = false}) async {
    final dto = SetPlayListOptionsRequestDto(loop: loop, shuffle: shuffle);
    await _invokeHubMethod('SetPlayListOptions', [id, dto]);
  }

  @override
  Future<void> deletePlayList(int id) async {
    await _invokeHubMethod('DeletePlayList', [id]);
  }

  @override
  Future<void> deleteFile(int id, int playListId) async {
    await _invokeHubMethod('DeleteFile', [playListId, id]);
  }

  @override
  Future<void> loopFile(int id, int playListId, {bool loop = false}) async {
    await _invokeHubMethod('LoopFile', [playListId, id, loop]);
  }

  @override
  Future<void> setFileOptions(int streamIndex, {bool isAudio = false, bool isSubtitle = false, bool isQuality = false}) async {
    final dto = SetFileOptionsRequestDto(streamIndex: streamIndex, isAudio: isAudio, isQuality: isQuality, isSubTitle: isSubtitle);
    await _invokeHubMethod('SetFileOptions', [dto]);
  }

  @override
  Future<void> updateSettings(ServerAppSettings dto) async {
    await _invokeHubMethod('UpdateSettings', [dto]);
  }

  @override
  Future<void> loadPlayLists() async {
    await _invokeHubMethod('SendPlayListsToClient', []);
  }

  @override
  Future<PlayListItemResponseDto> loadPlayList(int playListId) async {
    final response = await _invokeHubMethod('GetPlayList', [playListId]);
    return PlayListItemResponseDto.fromJson(response as Map<String, dynamic>);
  }

  @override
  Future<void> setVolume(double volumeLvl, {bool isMuted = false}) async {
    final dto = SetVolumeRequestDto(volumeLevel: volumeLvl, isMuted: isMuted);
    await _invokeHubMethod('SetVolume', [dto]);
  }

  @override
  Future<void> updatePlayList(int id, String name) async {
    final dto = UpdatePlayListRequestDto(name: name);
    await _invokeHubMethod('UpdatePlayList', [id, dto]);
  }
}
